From 24f0766d5b2b9ce69ac51de381c612162ab50df5 Mon Sep 17 00:00:00 2001
From: Leon Leijssen <leon.leijssen@topic.nl>
Date: Wed, 15 Apr 2020 22:33:20 +0530
Subject: [PATCH 1353/1852] mtd: spi-nor: Fix spi_nor_read for multi die flash
 connected in dual parallel mode

commit 50b6c3dc609a45d31e5116cf4a4ef39b8c858d27 from
https://github.com/Xilinx/linux-xlnx.git

When multi die flash chip(N25Q512) are connected in dual parallel mode.
flashcp fails at 64MB offset(so the 32MB boundary of the single chip).

In spi_nor_read a multiply of 2 was missing in the remaining page length
calculation which resulted in data corruption.

This patch updated spi_nor_read and adds the multiply by 2 in the remaining
page length calculation to resolve the issue.

Signed-off-by: Amit Kumar Mahapatra <amit.kumar-mahapatra@xilinx.com>
Signed-off-by: Leon Leijssen <leon.leijssen@topic.nl>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/mtd/spi-nor/spi-nor.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index 985e4b4b37c4..7271429fcda3 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -2964,9 +2964,9 @@ static int spi_nor_read(struct mtd_info *mtd, loff_t from, size_t len,
 				cur_bank = offset / bank_size;
 				nxt_bank = (offset + len) / bank_size;
 				if (cur_bank != nxt_bank)
-					rem_bank_len = (bank_size *
+					rem_bank_len = ((bank_size *
 							(cur_bank + 1)) -
-							offset;
+							offset) << nor->shift;
 				else
 					rem_bank_len = (mtd->size >>
 							stack_shift) -
-- 
2.31.1

